File: scripts/minc2volume-viewer.js
#!/usr/bin/env node
/*
* BrainBrowser: Web-based Neurological Visualization Tools
* (https://brainbrowser.cbrain.mcgill.ca)
*
* Copyright (C) 2011 McGill University 
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Affero General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
* Author: Tarek Sherif <tsherif@gmail.com> (http://tareksherif.ca/)
* Author: Nicolas Kassis
* Author: Paul Mougel
* Author: Robert D. Vincent <robert.d.vincent@mcgill.ca>
*/

/*
* Usage:
*   $ node minc2volume-viewer.js <filename>
* OR as a node module:
*   var minc2volumeViewer = require('./minc2volume-viewer')
*   minc2volumeViewer.getHeader('<filename>', function (err, header) {
*     ...
*   })
*   var rawStream = minc2volumeViewer.getRawDataStream('<filename>')
*   rawStream.pipe(...)
*   rawStream.on('error', function (err) { ... })
*/


"use strict";

var fs = require("fs");
var exec =  require("child_process").exec;
var spawn = require("child_process").spawn;
var version = "0.3.2";

var datatypes = [
  'int8', 'uint8', 'int16', 'uint16', 'int32', 'uint32', 'float32', 'float64'
];

if (require.main === module) {
  var args = process.argv.slice(2);
  var arg;
  while ((arg = args.shift()) !== undefined) {
    var filename;
    var datatype;
    if (arg === '-T') {
      datatype = args.shift();
      if (datatypes.indexOf(datatype) < 0) {
        printUsage();
        process.exit(0);
      }
    }
    else {
      filename = arg;
    }
  }

  if (filename === undefined) {
    printUsage();
    process.exit(0);
  }
  if (datatype === undefined) {
    datatype = 'uint8';
  }

  fs.exists(filename, function(exists) {

    if (!exists) {
      console.error("File " + filename + " does not exist.");
      process.exit(1);
    }

    fs.stat(filename, function(err, stat) {
      if (!stat.isFile()) {
        console.error(filename + " is not a valid file.");
        process.exit(1);
      }

      var basename = filename.match(/[^\/]+$/)[0];
      var rawDataStream, rawFileStream;

      console.log("Processing file:", filename);

      console.log("Creating header file: ", basename + ".header");
      getHeaderWithType(filename, datatype, function(err, header) {
        if (err) {
          return logExecutionError(err);
        }
        fs.writeFile(basename + ".header", JSON.stringify(header));
      });

      console.log("Creating raw data file: ", basename + ".raw");
      rawDataStream = getRawDataStream(filename, datatype);
      rawDataStream.on('error', logExecutionError);
      rawFileStream = fs.createWriteStream(basename + ".raw");
      rawDataStream.pipe(rawFileStream);
    });

  });
} else {
  module.exports.getHeader = getHeader;
  module.exports.getHeaderWithType = getHeaderWithType;
  module.exports.getRawDataStream = getRawDataStream;
}

///////////////////////////
// Helper functions
///////////////////////////

function printUsage() {
  console.log("minc2volume-viewer.js v" + version);
  console.log("Usage: node minc2volume-viewer.js [-T <datatype>] <filename>");
  console.log("       Where datatype is one of: ");
  for (var k in datatypes)
    console.log("           " + datatypes[k]);
}

function getRawDataStream(filename, datatype) {
  var args = ["-normalize", filename];

  if (datatype === undefined)
    datatype = 'uint8';

  switch (datatype) {
  case 'int8':
    args.unshift('-byte', '-signed');
    break;
  case 'uint8':
    args.unshift('-byte', '-unsigned');
    break;
  case 'int16':
    args.unshift('-short', '-signed');
    break;
  case 'uint16':
    args.unshift('-short', '-unsigned');
    break;
  case 'int32':
    args.unshift('-int', '-signed');
    break;
  case 'uint32':
    args.unshift('-int', '-unsigned');
    break;
  case 'float32':
    args.unshift('-float');
    break;
  case 'float64':
    args.unshift('-double');
    break;
  }
  var minctoraw = spawn("minctoraw", args);
  minctoraw.on("exit", function (code) {
    if (code === null || code !== 0) {
      var err = new Error("Process minctoraw failed with error code " + code);
      minctoraw.stdout.emit("error", err);
    }
  });
  minctoraw.on("error", function (err) {
    minctoraw.stdout.emit("error", err);
  });
  return minctoraw.stdout;
}

function getHeader(filename, callback) {
  getHeaderWithType(filename, 'uint8', callback);
}

function getHeaderWithType(filename, datatype, callback) {

  function getSpace(filename, header, space, callback) {
    header[space] = {};
    exec("mincinfo -attval " + space + ":start " + filename, function(error, stdout) {
      if (error) {
        return callback(error);
      }

      header[space].start = parseFloat(stdout);
      exec("mincinfo -dimlength " + space + " " + filename, function(error, stdout) {
        if (error) {
          return callback(error);
        }

        header[space].space_length = parseFloat(stdout);
        
        exec("mincinfo -attval " + space + ":step " + filename, function(error, stdout) {
          if (error) {
            return callback(error);
          }

          header[space].step = parseFloat(stdout);

          exec("mincinfo -attval " + space + ":direction_cosines " + filename, function(error, stdout) {
            if (error) {
              return callback(error);
            }

            var direction_cosines = stdout.replace(/^\s+/, "").replace(/\s+$/, "").split(/\s+/);

            if (direction_cosines.length > 1) {
              header[space].direction_cosines = direction_cosines.map(parseFloat);
            }

            callback(null, header);
          });
        });

      });

    });
  }

  function getOrder(header, filename, callback) {
    
    function handleOrder(order, callback) {
      header.order = order;
      
      if (order.length === 4) {
        exec("mincinfo -attval time:start " + filename, function(error, time_start) {
          if (error) {
            return callback(error);
          }

          exec("mincinfo -dimlength time " + filename, function(error, time_length) {
            if (error) {
              return callback(error);
            }

            header.time = {
              start: parseFloat(time_start),
              space_length: parseFloat(time_length)
            };
            callback(null, header);
          });
        });
      
      } else {
        callback(null, header);
      }
    
    }
    
    exec("mincinfo -attval image:dimorder " + filename, function(error, stdout) {
      var order = [];
      order = stdout.trim().split(",");
      
      if (order.length < 3 || order.length > 4) {
        exec("mincinfo -dimnames " + filename,function(error, stdout) {
          if (error) {
            return callback(error);
          }

          order = stdout.trim().split(" ");
          handleOrder(order, callback);
        });
      } else {
        handleOrder(order, callback);
      }
    
    });
  }

  function buildHeader(filename, datatype, callback) {
    var header = {};

    header.datatype = datatype;
   
    getOrder(header, filename, function(err, header) {
      if (err) {
        return callback(err);
      }
      getSpace(filename, header, "xspace", function(err, header) {
        if (err) {
          return callback(err);
        }
        getSpace(filename, header, "yspace", function(err, header) {
          if (err) {
            return callback(err);
          }
          getSpace(filename, header, "zspace",function(err, header) {
            if (err) {
              return callback(err);
            }
            if(header.order > 3) {
              getSpace(filename, header, "time", callback);
            } else {
              callback(null, header);
            }
          });
        });
      });
    });
  }

  buildHeader(filename, datatype, callback);
}

function logExecutionError(error) {
  error = error.toString();
  if (error.match("not installed") || error.match("command not found")) {
    console.error("\nminc2volume-viewer.js requires that the MINC tools be installed.");
    console.error("Visit http://www.bic.mni.mcgill.ca/ServicesSoftware/MINC for more info.\n");
  } else {
    console.trace(error);
  }
}
    
           

File: scripts/minc2volume-viewer.py
#!/usr/bin/env python
#
# BrainBrowser: Web-based Neurological Visualization Tools
# (https://brainbrowser.cbrain.mcgill.ca)
#
# Copyright (C) 2011 McGill University
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Author: Jon Pipitone <jon@pipitone.ca>
# Author: Robert D. Vincent <robert.d.vincent@mcgill.ca>
#         (Made it work with Python 2.7, supported multiple voxel types)

from __future__ import print_function
import sys
import shutil
import os.path
import argparse
import subprocess
import json

required_minc_cmdline_tools = ['mincinfo', 'minctoraw']

def console_error(message, exit_code):
    print(message, file=sys.stderr)
    sys.exit(exit_code)

def console_log(message):
    print(message)

def which(pgm):
    path=os.getenv('PATH')
    for p in path.split(os.path.pathsep):
        p=os.path.join(p, pgm)
        if os.path.exists(p) and os.access(p, os.X_OK):
            return p

def check_minc_tools_installed():
    for tool in required_minc_cmdline_tools:
        if not which(tool):
            console_error(
              "minc2volume-viewer.py requires that the MINC tools be installed.\n"
              "Visit http://www.bic.mni.mcgill.ca/ServicesSoftware/MINC for info.", 1)

def cmd(command):
    return subprocess.check_output(command.split(), universal_newlines=True).strip()

def get_space(mincfile, space):
    header = {
        "start" : float(cmd("mincinfo -attval {}:start {}".format(space,mincfile))),
        "space_length" : float(cmd("mincinfo -dimlength {} {}".format(space,mincfile))),
        "step" : float(cmd("mincinfo -attval {}:step {}".format(space,mincfile))),
    }

    cosines = cmd("mincinfo -attval {}:direction_cosines {}".format(space,mincfile))
    cosines = cosines.strip().split()
    if len(cosines) > 1:
        header["direction_cosines"] = list(map(float, cosines))

    return header

def make_header(mincfile, datatype, headerfile):
    header = {}

    header["datatype"] = datatype;

    # find dimension order
    order = cmd("mincinfo -attval image:dimorder {}".format(mincfile))
    order = order.split(",")

    if len(order) < 3 or len(order) > 4:
        order = cmd("mincinfo -dimnames {}".format(mincfile))
        order = order.split(" ")

    header["order"] = order

    if len(order) == 4:
        time_start  = cmd("mincinfo -attval time:start {}".format(mincfile))
        time_length = cmd("mincinfo -dimlength time {}".format(mincfile))

        header["time"] = { "start" : float(time_start),
                           "space_length" : float(time_length) }

    # find space
    header["xspace"] = get_space(mincfile,"xspace")
    header["yspace"] = get_space(mincfile,"yspace")
    header["zspace"] = get_space(mincfile,"zspace")

    if len(order) > 3:
        header["time"] = get_space(mincfile,"time")

    # write out the header
    open(headerfile,"w").write(json.dumps(header))

def make_raw(mincfile, datatype, rawfile):
    raw = open(rawfile, "wb")
    args = ["-normalize", mincfile]
    opts = {
        "int8":   ["-byte",  "-signed"],
        "uint8":  ["-byte",  "-unsigned"],
        "int16":  ["-short", "-signed"],
        "uint16": ["-short", "-unsigned"],
        "int32":  ["-int",   "-signed"],
        "uint32": ["-int",   "-unsigned"],
        "int32":  ["-int",   "-signed"],
        "uint32": ["-int",   "-unsigned"],
        "float32":["-float"],
        "float64":["-double"],
    }[datatype]
    for opt in opts:
        args.insert(0, opt)
    args.insert(0, "minctoraw")
    raw.write(
        subprocess.check_output(args))

def main(filename, datatype):
    if not os.path.isfile(filename):
        console_error("File {} does not exist.".format(filename), 1)

    check_minc_tools_installed()

    basename = os.path.basename(filename)
    headername = "{}.header".format(basename)
    rawname = "{}.raw".format(basename)

    console_log("Processing file: {}".format(filename))

    console_log("Creating header file: {}".format(headername))
    make_header(filename, datatype, headername)

    console_log("Creating raw data file: {}".format(rawname))
    make_raw(filename, datatype, rawname)

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("filename", help="the minc file to convert")
    parser.add_argument("-T", dest="datatype",
                        choices=["int8", "int16", "int32",
                                 "uint8", "uint16", "uint32",
                                 "float32", "float64"],
                        action="store", default="uint8",
                        help="set the voxel data type of the output")
    args = parser.parse_args()
    main(args.filename, args.datatype)
